### PR TITLE
ferdium: init at 6.0.0-nightly.40

### PR DESCRIPTION
**Pull request by magnouvean**
###### Description of changes

Ferdium is a recent fork of ferdi (which again is a fork of Franz) which has become quite popular. Like ferdi it keeps web applications in one place. See: https://ferdium.org/ for more info about the app.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by kciredor** [comment]
Now that Ferdi is dead this could really use a merge :100:

**Review by ericdallo — APPROVED** [review]
Can anyone please merge this?
@SuperSandro2000

**Comment by ericdallo** [comment]
@thiagokokada could you help merge this, please?

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -5967,6 +5967,10 @@ with pkgs;
     mkFranzDerivation = callPackage ../applications/networking/instant-messengers/franz/generic.nix { };
   };
 
+  ferdium = callPackage ../applications/networking/instant-messengers/ferdium {
+    mkFranzDerivation = callPackage ../applications/networking/instant-messengers/franz/generic.nix { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:5971]: Why is this not part of top-level?

File: pkgs/applications/networking/instant-messengers/ferdium/default.nix [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+          exec ${xdg-utils}/bin/xdg-open "$@"
+        ''))
+      ];
+    };
+  };
+in
+mkFranzDerivation' rec {
+  pname = "ferdium";
+  name = "Ferdium";
+  version = "6.0.0-nightly.40";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ferdium/default.nix:20]:
```suggestion
  pname = "ferdium";
  version = "6.0.0-nightly.40";
```

File: pkgs/applications/networking/instant-messengers/ferdium/default.nix [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  src = fetchurl {
+    url = "https://github.com/ferdium/ferdium-app/releases/download/v6.0.0-nightly.40/ferdium_6.0.0-nightly.40_amd64.deb";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ferdium/default.nix:22]: Please subsitute the version

File: pkgs/applications/networking/instant-messengers/ferdium/default.nix [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    sha256 = "sha256-pDVWVUQRJqFyaMGIhMekNzz8DVq+B8M3fD0JC8+7x2E=";
+  };
+  extraBuildInputs = [ xorg.libxshmfence ];
+  meta = with lib; {
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ferdium/default.nix:26]:
```suggestion
  };

  extraBuildInputs = [ xorg.libxshmfence ];

  meta = with lib; {
```

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (magnouvean) [reply pkgs/top-level/all-packages.nix:5971]: I more or less copied everything from the ferdi/franz configs (seeing as these programs are very similar) which seem to do it this way. If we make it part of top level we probably should change these packages as well?

Comment (thiagokokada) [reply pkgs/top-level/all-packages.nix]: I concur with @SuperSandro2000. We can probably expose this to top-level and use in both `ferdium` and `franz`.

**Comment by magnouvean** [comment]
It appears the name attribute is required for mkFranzDerivation, so I had to re-add this.

File: pkgs/applications/networking/instant-messengers/ferdium/default.nix [hunk 5]
```diff
@@ -0,0 +1,36 @@
+{ lib, mkFranzDerivation, fetchurl, xorg, xdg-utils, buildEnv, writeShellScriptBin }:
+
+let
+  mkFranzDerivation' = mkFranzDerivation.override {
+    xdg-utils = buildEnv {
+      name = "xdg-utils-for-ferdi";
+      paths = [
+        xdg-utils
+        (lib.hiPrio (writeShellScriptBin "xdg-open" ''
+          unset GDK_BACKEND
```
Comment (thiagokokada) [inline pkgs/applications/networking/instant-messengers/ferdium/default.nix:10]: Why is this necessary? Looks to me that instead of unsetting here, you should instead wrapper the `ferdium` derivation instead.

Comment (magnouvean) [reply pkgs/applications/networking/instant-messengers/ferdium/default.nix:10]: This is another thing I just copied over from the ferdi package, but can't really see why it is there, though I assume the ferdi packagers did this for a reason. When I remove the whole `mkFranzDerivation'` and replace it with the normal `mkFranzDerivation` the package still builds fine. Should I remove it then?

Comment (thiagokokada) [reply pkgs/applications/networking/instant-messengers/ferdium/default.nix:10]: If this is also done in the Franz derivation you can probably just add this directly to the `mkFranzDerivation` function instead of duplicating the code here.

Comment (magnouvean) [reply pkgs/applications/networking/instant-messengers/ferdium/default.nix:10]: It doesn't seem to be in `mkFranzDerivation`. The binary seems to work fine with `GDK_BACKEND=wayland` and `GDK_BACKEND=x11` (which is what this variable should be) and runs fine in wayland. I'll just remove the whole `mkFranzDerivation'` and move to `mkFranzDerivation` it and if someone has any problem with it they can open an issue.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix]: > I more or less copied everything from the ferdi/franz configs (seeing as these programs are very similar) which seem to do it this way. If we make it part of top level we probably should change these packages as well?

yes, we don't need and shouldn't do copy pasta this all the time.

**Comment by SuperSandro2000** [comment]
@ofborg build ferdium

**Comment by ericdallo** [comment]
Thank you @SuperSandro2000 !
